Kernel/Net/VirtIO/VirtIONetworkAdapter.cpp

```diff
@@ -0,0 +1,283 @@
+    LittleEndian<u16> num_buffers;
+    u8 frame[0];
+};
+
+}
+
+using namespace VirtIO;
+
+static constexpr u16 RECEIVEQ = 0;
+static constexpr u16 TRANSMITQ = 1;
+
+static constexpr size_t MAX_RX_FRAME_SIZE = 1514; // Non-jumbo Ethernet frame limit.
+static constexpr size_t RX_BUFFER_SIZE = sizeof(VirtIONetHdr) * MAX_RX_FRAME_SIZE;
+static constexpr u16 MAX_INFLIGHT_PACKETS = 128;
+
+UNMAP_AFTER_INIT ErrorOr<bool> VirtIONetworkAdapter::probe(PCI::DeviceIdentifier const& pci_device_identifier)
+{
+    if (pci_device_identifier.hardware_id().vendor_id != PCI::VendorID::VirtIO)
+        return false;
+    if (pci_device_identifier.hardware_id().device_id != PCI::DeviceID::VirtIONetAdapter)
+        return false;
+    return true;
+}
+
+UNMAP_AFTER_INIT ErrorOr<NonnullRefPtr<NetworkAdapter>> VirtIONetworkAdapter::create(PCI::DeviceIdentifier const& pci_device_identifier)
+{
+    auto interface_name = TRY(NetworkingManagement::generate_interface_name_from_pci_address(pci_device_identifier));
+    return TRY(adopt_nonnull_ref_or_enomem(new (nothrow) VirtIONetworkAdapter(pci_device_identifier, move(interface_name))));
+}
+
+UNMAP_AFTER_INIT VirtIONetworkAdapter::VirtIONetworkAdapter(PCI::DeviceIdentifier const& pci_device_identifier, NonnullOwnPtr<KString> interface_name)
+    : VirtIO::Device(pci_device_identifier)
+    , NetworkAdapter(move(interface_name))
+{
+}
+
+UNMAP_AFTER_INIT ErrorOr<void> VirtIONetworkAdapter::initialize(Badge<NetworkingManagement>)
+{
+    m_rx_buffers = TRY(Memory::RingBuffer::try_create("VirtIONetworkAdapter Rx buffer"sv, RX_BUFFER_SIZE * MAX_INFLIGHT_PACKETS));
+    m_tx_buffers = TRY(Memory::RingBuffer::try_create("VirtIONetworkAdapter Tx buffer"sv, RX_BUFFER_SIZE * MAX_INFLIGHT_PACKETS));
+
+    return initialize_virtio_resources();
+}
+
+UNMAP_AFTER_INIT ErrorOr<void> VirtIONetworkAdapter::initialize_virtio_resources()
+{
+    dbgln_if(VIRTIO_DEBUG, "VirtIONetworkAdapter: initialize_virtio_resources");
+    TRY(Device::initialize_virtio_resources());
+    m_device_config = TRY(get_config(VirtIO::ConfigurationType::Device));
+
+    bool success = negotiate_features([&](u64 supported_features) {
+        u64 negotiated = 0;
+        if (is_feature_set(supported_features, VIRTIO_NET_F_STATUS))
+            negotiated |= VIRTIO_NET_F_STATUS;
+        if (is_feature_set(supported_features, VIRTIO_NET_F_MAC))
+            negotiated |= VIRTIO_NET_F_MAC;
+        if (is_feature_set(supported_features, VIRTIO_NET_F_SPEED_DUPLEX))
+            negotiated |= VIRTIO_NET_F_SPEED_DUPLEX;
+        if (is_feature_set(supported_features, VIRTIO_NET_F_MTU))
+            negotiated |= VIRTIO_NET_F_MTU;
+        return negotiated;
+    });
+    if (!success)
+        return Error::from_errno(EIO);
+
+    success = handle_device_config_change();
+    if (!success)
+        return Error::from_errno(EIO);
+
+    success = setup_queues(2); // receive & transmit
+    if (!success)
+        return Error::from_errno(EIO);
```

Comment:
This pattern is quite meh. A follow-up change to update the base class's virtual methods here to return `ErrorOr<void>` would be a nice Kernel refactoring PR. That way we could just TRY() these methods.